ext/opentelemetry-ext-wsgi/src/opentelemetry/ext/wsgi/__init__.py

```diff
@@ -88,14 +118,27 @@ def __call__(self, environ, start_response):
         tracer = trace.tracer()
         path_info = environ["PATH_INFO"] or "/"
 
-        with tracer.start_span(path_info, kind=trace.SpanKind.SERVER) as span:
+        span = tracer.create_span(path_info, kind=trace.SpanKind.SERVER)
+        span.start()
+        try:
             self._add_request_attributes(span, environ)
             start_response = self._create_start_response(span, start_response)
 
             iterable = self.wsgi(environ, start_response)
-            try:
-                for yielded in iterable:
-                    yield yielded
-            finally:
-                if hasattr(iterable, "close"):
-                    iterable.close()
+
+            # Put this in a subfunction to not delay the call to the wrapped
+            # WSGI application (instrumentation should change the application
+            # behavior as little as possible).
+            def iter_result():
+                try:
+                    for yielded in iterable:
+                        yield yielded
+                finally:
+                    if hasattr(iterable, "close"):
+                        iterable.close()
+                    span.end()
```

Comment:
Rather than defining this as a closure, it should be more efficient to define this as a function which takes an `iterable` argument.

```python
def iter_result(span, iterable):
    try:
        for yielded in iterable:
            yield yielded
    finally:
        if hasattr(iterable, "close"):
            iterable.close()
        span.end()
```

Reply:
Have you measured it though? In theory, creating the closure could be cheaper than passing an argument.

Reply:
Seems it is faster indeed:
```python
In [1]: def mkiter():
   ...:     xs = ('',)
   ...:     def doiter():
   ...:         for x in xs:
   ...:             yield x
   ...:     return doiter()
   ...:
In [8]: def mkiter2():
   ...:     xs = ('',)
   ...:     def doiter2(xs_):
   ...:         for x in xs_:
   ...:             yield x
   ...:     return doiter2(xs)
   ...:

In [9]: %timeit for _ in mkiter2(): pass
287 ns ± 2.38 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [10]: %timeit for _ in mkiter(): pass
339 ns ± 3.49 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
I meant that it should be faster to define `iter_result` as purely a global rather than defining it from within the `__call__`. Maybe even as a `staticmethod` on the middleware, but probably just easiest to define this as a global. Both mechanisms in the example above are still closures.

Reply:
So:
```python
class OpenTelemetryMiddleware:
  @staticmethod
  def iter_result(span, iterable):
    ...
  def __call__(self, environ, start_response):
    ...
    return self.iter_result(span, iterable)
```

or

```python
def iter_result(span, iterable):
    ...

class OpenTelemetryMiddleware:
  def __call__(self, environ, start_response):
    ...
   return iter_result(span, iterable)
```

Reply:
I think defining it as a global might even be slower because it requires a look-up in the globals-dictionary. Although on the other hand, the function object doesn't have to be created each time. We are getting in micro-optimization territory here, but I'll look into it anyway.

Reply:
> Both mechanisms in the example above are still closures.

I don't know your exact definition of closure, but at least the second example has the property `dotiter2.__closure__ is None`:

```python
In [16]: def mkiter():
    ...:     xs = ('',)
    ...:     def doiter():
    ...:         for x in xs:
    ...:             yield x
    ...:     print(doiter.__closure__)
    ...:     return doiter()
    ...:
    ...: def mkiter2():
    ...:     xs = ('',)
    ...:     def doiter2(xs_):
    ...:         for x in xs_:
    ...:             yield x
    ...:     print(doiter2.__closure__)
    ...:     return doiter2(xs)
    ...:

In [19]: mkiter()
(<cell at 0x00000200461D5FD8: tuple object at 0x0000020046253160>,)
Out[19]: <generator object mkiter.<locals>.doiter at 0x0000020046264A20>

In [20]: mkiter2()
None
Out[20]: <generator object mkiter2.<locals>.doiter2 at 0x0000020046264318>
```

Reply:
OK, so it seems this is indeed still a bit faster (but only very slightly):

```python
In [31]: def mkiter():
    ...:     xs = ('',)
    ...:     def doiter():
    ...:         for x in xs:
    ...:             yield x
    ...:     return doiter()
    ...:
    ...: def mkiter2():
    ...:     xs = ('',)
    ...:     def doiter2(xs_):
    ...:         for x in xs_:
    ...:             yield x
    ...:     return doiter2(xs)
    ...:
    ...: def doiter3(xs_):
    ...:     for x in xs_:
    ...:         yield x
    ...:
    ...: def mkiter3():
    ...:     xs = ('',)
    ...:     return doiter3(xs)
    ...:
    ...: def mkiter4(_doiter=doiter3):
    ...:     xs = ('',)
    ...:     return _doiter(xs)
    ...:

In [32]:  %timeit for _ in mkiter(): pass
346 ns ± 31.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [33]:  %timeit for _ in mkiter2(): pass
287 ns ± 5.75 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [34]:  %timeit for _ in mkiter3(): pass
259 ns ± 5.75 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [35]:  %timeit for _ in mkiter4(): pass
256 ns ± 13.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
I'm not changing this because it seems looking up a static method via self is slow enough to make the improvement go away. Unless you also want to move all other static methods to globals or optimize access to them. This has become too much of a micro-optimization for my taste:

```python
In [38]: class WSGIMiddleWare:
    ...:     @staticmethod
    ...:     def doiter5(xs_):
    ...:         for x in xs_:
    ...:             yield x
    ...:
    ...:     def mkiter5(self):
    ...:         xs = ('',)
    ...:         return self.doiter5(xs)
    ...:

In [41]: mkiter5a = WSGIMiddleWare().mkiter5

In [42]: %timeit for _ in mkiter5a(): pass
286 ns ± 11.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
I moved it to a global after all, because it also reduces nesting and is thus IMHO better readable.